File: code/backend/app/schemas/mt.py
"""健康数据中台模块 Schema（技术架构第11.2章）。"""
from datetime import date, datetime, time
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict

from app.schemas.common import BaseOut


# ---------------- 客户 ----------------
class CustomerCreate(BaseModel):
    name_mask: Optional[str] = None
    gender: Optional[str] = None
    birth_date: Optional[datetime] = None
    phone_mask: Optional[str] = None
    source_store_id: Optional[int] = None
    health_tags: Optional[dict] = None
    auth_status: str = "unauthorized"
    auth_file_url: Optional[str] = None


class CustomerAuthorizeIn(BaseModel):
    auth_file_url: Optional[str] = None


class CustomerOut(BaseOut):
    name_mask: Optional[str] = None
    gender: Optional[str] = None
    birth_date: Optional[datetime] = None
    phone_mask: Optional[str] = None
    source_store_id: Optional[int] = None
    health_tags: Optional[dict] = None
    auth_status: str


# ---------------- 疼痛评估 ----------------
class PainAssessmentCreate(BaseModel):
    customer_id: int
    assess_time: Optional[datetime] = None
    scale_type: Optional[str] = None
    answers_json: Optional[dict] = None
    score: Optional[float] = None
    pain_site: Optional[str] = None
    pain_nature: Optional[str] = None
    therapist_id: Optional[int] = None


class PainAssessmentOut(BaseOut):
    customer_id: int
    assess_time: Optional[datetime] = None
    scale_type: Optional[str] = None
    answers_json: Optional[dict] = None
    score: Optional[float] = None
    pain_site: Optional[str] = None
    pain_nature: Optional[str] = None
    therapist_id: Optional[int] = None


# ---------------- 调理方案 ----------------
class CarePlanCreate(BaseModel):
    customer_id: int
    source_store_id: Optional[int] = None
    created_by: Optional[int] = None
    doctor_advice_id: int  # 强制关联执业医师出具的方案建议/处方 ID（合规强规则1）
    pain_type: Optional[str] = None
    goal: Optional[str] = None
    cycle: Optional[str] = None
    age: Optional[int] = None
    pain_score: Optional[int] = None
    chronic_count: Optional[int] = None


class CarePlanOut(BaseOut):
    customer_id: int
    doctor_advice_id: Optional[int] = None
    pain_type: Optional[str] = None
    goal: Optional[str] = None
    cycle: Optional[str] = None
    items_json: Any = None
    product_combo_json: Any = None
    reeval_nodes: Any = None
    status: str


# ---------------- 治疗记录 ----------------
class TreatmentRecordCreate(BaseModel):
    customer_id: int
    store_id: int
    therapist_id: Optional[int] = None
    plan_id: Optional[int] = None
    service_time: Optional[datetime] = None
    products_json: Optional[dict] = None
    oper_sites_json: Optional[dict] = None
    nps: Optional[int] = None
    images_json: Optional[dict] = None
    remark: Optional[str] = None


class TreatmentRecordOut(BaseOut):
    customer_id: int
    store_id: int
    therapist_id: Optional[int] = None
    plan_id: Optional[int] = None
    service_time: Optional[datetime] = None
    products_json: Optional[dict] = None
    oper_sites_json: Optional[dict] = None
    nps: Optional[int] = None
    images_json: Optional[dict] = None
    remark: Optional[str] = None


class TreatmentRecordReviseIn(BaseModel):
    """治疗记录更正（合规强规则2：不可删，仅可更正留痕）。"""
    products_json: Optional[dict] = None
    oper_sites_json: Optional[dict] = None
    nps: Optional[int] = None
    images_json: Optional[dict] = None
    remark: Optional[str] = None
    reason: Optional[str] = None  # 更正事由，留痕必填建议


class TreatmentRecordRevisionOut(BaseOut):
    record_id: int
    revised_by: Optional[int] = None
    revised_by_role: Optional[str] = None
    before_json: Any = None
    after_json: Any = None
    reason: Optional[str] = None


# ---------------- 门店 / 调理师 ----------------
class StoreOut(BaseOut):
    name: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    type: Optional[str] = None
    status: str


class TherapistOut(BaseOut):
    name: Optional[str] = None
    license_no: Optional[str] = None
    store_id: int
    skill_tags: Optional[dict] = None
    status: str


# ---------------- 复购预测（AI 反馈闭环，第15.4章） ----------------
class RepurchasePredictIn(BaseModel):
    customer_id: int
    age: int = 40
    visit_freq: float = 3.0
    last_gap_days: float = 30.0


class RepurchasePredictionOut(BaseOut):
    customer_id: int
    predict_time: Optional[datetime] = None
    next_visit_prob: Optional[float] = None
    repurchase_prob: Optional[float] = None
    risk_level: Optional[str] = None
    model_version: Optional[str] = None


# ---------------- 风险画像（AI 反馈闭环，第15.4章） ----------------
class RiskProfileIn(BaseModel):
    customer_id: int
    age: int = 45
    bmi: float = 24.0
    comorbidity_count: int = 0


class RiskProfileOut(BaseOut):
    customer_id: int
    predict_time: Optional[datetime] = None
    pain_risk: Optional[str] = None
    comorbidity_risk: Optional[str] = None
    model_version: Optional[str] = None


# ---------------- 门店经营宽表（第11.3章） ----------------
class StoreMetricsOut(BaseOut):
    date: Optional[str] = None
    store_id: Optional[int] = None
    store_name: Optional[str] = None
    region: Optional[str] = None
    appointment_cnt: int = 0
    arrival_cnt: int = 0
    deal_customers: int = 0
    deal_amount: float = 0.0
    deal_orders: int = 0
    repurchase_customers: int = 0
    nps_avg: float = 0.0


class AggregateMetricsIn(BaseModel):
    target_date: date
    store_id: Optional[int] = None


# ---------------- P6: 调理师排班 ----------------
class TherapistScheduleCreate(BaseModel):
    therapist_id: int
    work_date: date
    am_pm: str = "morning"  # morning/afternoon/evening
    start_time: time
    end_time: time
    capacity: int = 1
    remark: Optional[str] = None


class TherapistScheduleUpdate(BaseModel):
    work_date: Optional[date] = None
    am_pm: Optional[str] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    status: Optional[str] = None  # open/closed
    capacity: Optional[int] = None
    remark: Optional[str] = None


class TherapistScheduleOut(BaseOut):
    therapist_id: int
    store_id: int
    work_date: date
    am_pm: str
    start_time: time
    end_time: time
    status: str
    capacity: int
    remark: Optional[str] = None


# ---------------- P6: 调理师标签 ----------------
class TherapistTagCreate(BaseModel):
    name: str
    category: Optional[str] = None  # skill/health/service
    description: Optional[str] = None


class TherapistTagOut(BaseOut):
    name: str
    category: Optional[str] = None
    description: Optional[str] = None


class TherapistTagRelOut(BaseModel):
    therapist_id: int
    tag_id: int
    tag_name: Optional[str] = None
    category: Optional[str] = None
    assigned_by: Optional[int] = None
    created_at: Optional[datetime] = None


# ---------------- 效果四档判定（合规强规则3，第11.2章） ----------------
class EffectTrackingCreate(BaseModel):
    customer_id: int
    plan_id: Optional[int] = None
    baseline_pain: Optional[float] = None  # 基线疼痛评分（方案开始时）
    latest_pain: Optional[float] = None    # 最近一次疼痛评分
    nps: Optional[int] = None
    repurchase_count: Optional[int] = None


class EffectTrackingOut(BaseOut):
    customer_id: int
    plan_id: Optional[int] = None
    effect_level: Optional[str] = None  # significant/effective/ineffective/worsened
    assess_seq_json: Any = None
    generated_at: Optional[datetime] = None

File: code/mp-taro/src/services/mt.ts
// 健康数据中台（mt）服务封装（第11.2/11.3章：客户/疼痛评估/照护计划/治疗记录/复购/风险）。
// 风格对齐 services/ih.ts：request + API 常量 + PageResult 泛型。
import { request } from './request'
import { API } from '@/constants/api'
import type { PageResult } from './ih'

// ---------------- 客户 ----------------
export interface Customer {
  id: number
  name_mask?: string
  gender?: string
  phone_mask?: string
  source_store_id?: number
  auth_status: string // unauthorized | authorized
  created_at?: string
}

export async function listCustomers(params: { page?: number; page_size?: number; auth_status?: string }) {
  return request<PageResult<Customer>>(API.customers, { data: params })
}

export async function createCustomer(body: {
  name_mask?: string
  gender?: string
  phone_mask?: string
  source_store_id?: number
  health_tags?: Record<string, unknown>
}) {
  return request<Customer>(API.customers, { method: 'POST', data: body })
}

export async function authorizeCustomer(customerId: number, body?: { auth_file_url?: string }) {
  return request<Customer>(`${API.customers}/${customerId}/authorize`, { method: 'PATCH', data: body || {} })
}

export async function getCustomer(customerId: number) {
  return request<Customer>(`${API.customers}/${customerId}`)
}

// ---------------- 疼痛评估 ----------------
export interface PainAssessment {
  id: number
  customer_id: number
  score?: number
  pain_site?: string
  pain_nature?: string
  scale_type?: string
  assess_time?: string
  created_at?: string
}

export async function listPainAssessments(params: { page?: number; page_size?: number; customer_id?: number }) {
  return request<PageResult<PainAssessment>>(API.painAssessments, { data: params })
}

export async function createPainAssessment(body: {
  customer_id: number
  score?: number
  pain_site?: string
  pain_nature?: string
  scale_type?: string
  answers_json?: Record<string, unknown>
}) {
  return request<PainAssessment>(API.painAssessments, { method: 'POST', data: body })
}

// ---------------- 照护计划 ----------------
export interface CarePlan {
  id: number
  customer_id: number
  doctor_advice_id?: number
  pain_type?: string
  goal?: string
  cycle?: string
  status: string
  items_json?: unknown
  product_combo_json?: unknown
  created_at?: string
}

export async function listCarePlans(params: { page?: number; page_size?: number; customer_id?: number }) {
  return request<PageResult<CarePlan>>(API.carePlans, { data: params })
}

export async function createCarePlan(body: {
  customer_id: number
  doctor_advice_id: number
  pain_type?: string
  goal?: string
  cycle?: string
  age?: number
  pain_score?: number
  chronic_count?: number
}) {
  return request<CarePlan>(API.carePlans, { method: 'POST', data: body })
}

// ---------------- 治疗记录 ----------------
export interface TreatmentRecord {
  id: number
  customer_id: number
  store_id: number
  therapist_id?: number
  plan_id?: number
  nps?: number
  service_time?: string
  remark?: string
  created_at?: string
}

export async function listTreatmentRecords(params: { page?: number; page_size?: number; customer_id?: number }) {
  return request<PageResult<TreatmentRecord>>(API.treatmentRecords, { data: params })
}

export async function createTreatmentRecord(body: {
  customer_id: number
  store_id: number
  therapist_id?: number
  plan_id?: number
  nps?: number
  products_json?: Record<string, unknown>
  remark?: string
}) {
  return request<TreatmentRecord>(API.treatmentRecords, { method: 'POST', data: body })
}

// 合规强规则2：治疗记录不可删，仅可更正留痕（PATCH 写入修订表 + 审计前后对照）
export async function reviseTreatmentRecord(
  recordId: number,
  body: {
    products_json?: Record<string, unknown>
    oper_sites_json?: Record<string, unknown>
    nps?: number
    images_json?: Record<string, unknown>
    remark?: string
    reason: string
  },
) {
  return request<TreatmentRecord>(`${API.treatmentRecords}/${recordId}`, { method: 'PATCH', data: body })
}

// ---------------- 门店 / 调理师 ----------------
export interface Store {
  id: number
  name?: string
  region?: string
  address?: string
}

export async function listStores(params: { page?: number; page_size?: number; region?: string }) {
  return request<PageResult<Store>>(API.stores, { data: params })
}

export interface Therapist {
  id: number
  store_id: number
  name_mask?: string
}

export async function listTherapists(storeId: number, params?: { page?: number; page_size?: number }) {
  return request<PageResult<Therapist>>(`${API.stores}/${storeId}/therapists`, { data: params })
}

// ---------------- 效果四档（合规强规则3） ----------------
export interface EffectTracking {
  id: number
  customer_id: number
  plan_id?: number
  effect_level?: string // significant | effective | ineffective | worsened
  generated_at?: string
}

export async function listEffectTracking(params: {
  page?: number
  page_size?: number
  customer_id?: number
  plan_id?: number
  effect_level?: string
}) {
  return request<PageResult<EffectTracking>>(API.effectTracking, { data: params })
}

export async function createEffectTracking(body: {
  customer_id: number
  plan_id?: number
  baseline_pain?: number
  latest_pain?: number
  nps?: number
  repurchase_count?: number
}) {
  return request<EffectTracking>(API.effectTracking, { method: 'POST', data: body })
}

// ---------------- 复购预测（AI 反馈闭环） ----------------
export interface RepurchasePrediction {
  id: number
  customer_id: number
  next_visit_prob?: number
  repurchase_prob?: number
  risk_level?: string
  model_version?: string
  predict_time?: string
}

export async function listRepurchasePredictions(params: { page?: number; page_size?: number; customer_id?: number }) {
  return request<PageResult<RepurchasePrediction>>(API.repurchasePredictions, { data: params })
}

export async function predictRepurchase(body: { customer_id: number; age?: number; visit_freq?: number; last_gap_days?: number }) {
  return request<RepurchasePrediction>(API.repurchasePredictions, { method: 'POST', data: body })
}

// ---------------- 风险画像（AI 反馈闭环） ----------------
export interface RiskProfile {
  id: number
  customer_id: number
  pain_risk?: string
  comorbidity_risk?: string
  model_version?: string
  predict_time?: string
}

export async function listRiskProfiles(params: { page?: number; page_size?: number; customer_id?: number }) {
  return request<PageResult<RiskProfile>>(API.riskProfiles, { data: params })
}

export async function predictRisk(body: { customer_id: number; age?: number; bmi?: number; comorbidity_count?: number }) {
  return request<RiskProfile>(API.riskProfiles, { method: 'POST', data: body })
}
